tests/conftest.py

```diff
@@ -34,4 +50,220 @@ def pytest_collection_modifyitems(config, items):
 
 def pytest_generate_tests(metafunc):
     if "seed" in metafunc.fixturenames:
-        metafunc.parametrize("seed", metafunc.config.getoption("seed"))
+        metafunc.parametrize("seed", metafunc.config.getoption("seed"), scope="session")
+
+## """Fixtures and classes that are to be utilised for sharing simulations across the test framework."""
+
+if TYPE_CHECKING:
+    from pandas import DataFrame
+
+@pytest.fixture(scope="session")
+def jan_1st_2010() -> Date:
+    return Date(2010, 1, 1)
+
+
+@pytest.fixture(scope="session")
+def resource_filepath() -> Path:
+    return (Path(os.path.dirname(__file__)) / "../resources").resolve()
+
+
+@pytest.fixture(scope="session")
+def small_shared_sim(seed, jan_1st_2010, resource_filepath):
+    """
+    Shared simulation object that can be re-used across multiple tests.
+
+    Note that to ensure the object can be shared between tests, it is
+    necessary that:
+    - All modules that are to make use of the shared simulation,
+    and their dependencies, are registered with the simulation.
+    - The initial population cannot be changed and must be sufficient
+    for all tests in which the object is to be used.
+    - The simulation cannot be further simulated into the future.
+    """
+    sim = Simulation(start_date=jan_1st_2010, seed=seed)
+    sim.register(
+        demography.Demography(resourcefilepath=resource_filepath),
+        diarrhoea.Diarrhoea(resourcefilepath=resource_filepath, do_checks=True),
+        diarrhoea.DiarrhoeaPropertiesOfOtherModules(),
+        enhanced_lifestyle.Lifestyle(resourcefilepath=resource_filepath),
+        healthburden.HealthBurden(resourcefilepath=resource_filepath),
+        healthseekingbehaviour.HealthSeekingBehaviour(
+            resourcefilepath=resource_filepath,
+            force_any_symptom_to_lead_to_healthcareseeking=True,
+        ),
+        healthsystem.HealthSystem(
+            resourcefilepath=resource_filepath, cons_availability="all"
+        ),
+        simplified_births.SimplifiedBirths(resourcefilepath=resource_filepath),
+        stunting.Stunting(resourcefilepath=resource_filepath),
+        symptommanager.SymptomManager(resourcefilepath=resource_filepath),
+    )
+    sim.make_initial_population(n=100)
+    sim.simulate(end_date=sim.start_date)
+    return sim
+
+
+class _BaseSharedSim:
```

Comment:
The idea of using fixtures to safely mediate updates to the shared simulation fixture is cool, I hadn't come across the use of `yield` in fixtures to allow setup / teardown before. 

What I'm less sure of is if we need the `_BaseSharedSim` class and corresponding subclasses? We could do something similar with just test functions, for example

```Python
@pytest.fixture(scope="session")
def start_date() -> Date:
    return Date(2010, 1, 1)

@pytest.fixture(scope="session")
def resource_filepath() -> Path:
    return (Path(os.path.dirname(__file__)) / "../resources").resolve()

@pytest.fixture(scope="session")
def shared_sim(seed, start_date, resource_filepath):
    ...

@pytest.fixture
def clear_shared_sim_hsi_event_queue(shared_sim):
    cached_queue = shared_sim_healthsystem.HSI_EVENT_QUEUE.copy()
    shared_sim.modules["HealthSystem"].reset_queue()
    yield
    shared_sim.modules["HealthSystem"].HSI_EVENT_QUEUE = cached_queue

@pytest.fixture
def changes_shared_sim_event_queue(shared_sim):
    cached_event_queue = shared_sim.event_queue.copy()
    yield
    shared_sim.event_queue = cached_event_queue

@pytest.fixture
def changes_shared_sim_module_parameters(shared_sim, changed_module_name):
    cached_parameter_values = shared_sim.modules[changed_module_name].parameters.copy()
    yield
    shared_sim.modules[changed_module_name] = cached_parameter_values 

@pytest.fixture
def changes_shared_sim_person_properties(shared_sim, changed_person_id):
    cached_values = shared_sim_df.loc[changed_person_id].copy()
    yield
    shared_sim_df.loc[changed_person_id] = cached_values
```

This assumes a `changed_module_name` / `changed_person_id` fixture would be defined in the scope of the test function using the `changes_shared_sim_module_parameters`  / `changes_shared_sim_person_properties` fixture, which would work when we want to use the same module / person ID values across a test module (or test class). If we wanted to also be able to set these at a per test function level we could use [the ability to use marks to pass data to fixtures ](https://docs.pytest.org/en/6.2.x/fixture.html#using-markers-to-pass-data-to-fixtures) and have defaults for the fixtures like

```Python
@pytest.fixture
def changed_module_name(request):
    return request.node.get_closest_marker("changed_module_name").args[0]

@pytest.fixture
def changed_person_id(request):
    return request.node.get_closest_marker("changed_person_id").args[0]
```

with an example test function usage then something like

```Python
@pytest.mark.changed_module_name("Diarrhoea")
@pytest.mark.changed_person_id(0)
def test_do_when_presentation_with_diarrhoea_severe_dehydration(
    shared_sim,
    changed_person_id,
    changes_shared_sim_module_parameters,
    changes_shared_sim_person_properties,
    clear_shared_sim_hsi_event_queue,
):
    ...
```

Reply:
Also if we do keep the base class approach, I would say we don't want to use an underscrore prefixed name as it is intended for use outside this module?